src/sage/modular/drinfeld_modform/ring.py

```diff
@@ -0,0 +1,653 @@
+
+        ..MATH::
+
+            \phi_{w, a} = a + g_{1, a}\tau + \cdots + g_{r d_a, a}\tau^{r d_a}
+
+        where `d_a := \mathrm{deg}(a)`.
+
+        INPUT:
+
+        - ``i`` -- an integer between 1 and `r d_a`;
+
+        - ``a`` -- (default: ``None``) an element in the ring of regular
+          functions. If `a` is ``None``, then the method returns the
+          `i`-th coefficient form of `\phi_{w, T}`.
+
+        EXAMPLES::
+
+            sage: q = 3
+            sage: A = GF(q)['T']
+            sage: K.<T> = Frac(A)
+            sage: M = DrinfeldModularForms(K, 3)
+            sage: M.coefficient_form(1)
+            g1
+            sage: M.coefficient_form(2)
+            g2
+            sage: M.coefficient_form(3)
+            g3
+            sage: M.coefficient_form(3, T^2)
+            g1^9*g2 + g1*g2^3 + (T^27 + T)*g3
+
+        ::
+
+            sage: M = DrinfeldModularForms(K, 2, has_type=True)
+            sage: M.coefficient_form(1)
+            g1
+            sage: M.coefficient_form(2)
+            h2^2
+            sage: M.coefficient_form(2, T^3 + T^2 + T)
+            (T^9 + T^3 + T + 1)*g1^4 + (T^18 + T^10 + T^9 + T^2 + T + 1)*h2^2
+
+        TESTS::
+
+            sage: M.coefficient_form(0)
+            Traceback (most recent call last):
+            ...
+            ValueError: index (=0) must be >= 1 and <= rank (=2)
+            sage: M.coefficient_form(3)
+            Traceback (most recent call last):
+            ...
+            ValueError: index (=3) must be >= 1 and <= rank (=2)
+            sage: M.coefficient_form(9, T^2)
+            Traceback (most recent call last):
+            ...
+            ValueError: index (=9) must be >= 1 and <= deg(a)*rank (=4)
+            sage: M.coefficient_form(1, 1/T)
+            Traceback (most recent call last):
+            ...
+            ValueError: a should be in the ring of regular functions
+            sage: M.coefficient_form(1, x)
+            Traceback (most recent call last):
+            ...
+            TypeError: a should be an element of the base ring
+        """
+        i = ZZ(i)
+        if a is None:
+            if i < 1 or i > self.rank():
+                raise ValueError(f"index (={i}) must be >= 1 and <= rank "
+                                 f"(={self.rank()})")
+            return self._generator_coefficient_form(i)
+        try:
+            a = self._base_ring(a)
+        except TypeError:
+            raise TypeError("a should be an element of the base ring")
+        if not a.denominator().is_one():
+            raise ValueError("a should be in the ring of regular functions")
+        a = a.numerator()
+        if i < 1 or i > a.degree()*self.rank():
+            raise ValueError(f"index (={i}) must be >= 1 and <= deg(a)*rank "
+                             f"(={a.degree()*self.rank()})")
+        coeff_forms = self._coefficient_forms(a)
+        return coeff_forms[i - 1]
+
+    def coefficient_forms(self, a=None):
```

Comment:
Why not removing `_coefficient_forms`, putting it in `coefficient_forms` and updating `coefficient_form` accordingly?

Reply:
The technicality here which led me to this implementation is because the following check in the method `coefficient_form`:
```python
if i < 1 or i > a.degree()*self.rank():
    raise ValueError(f"index (={i}) must be >= 1 and <= deg(a)*rank "
                     f"(={a.degree()*self.rank()})")
``` 
is necessary as I need to get the degree of `a` in order to check if the index `i` is in the correct range. So calling the method `coefficient_forms` (plural) would do a redundant check.

Anyway, maybe I'm missing a point here so I'm open if you have a cleaner solution.

Reply:
Okay, fair enough.